File: packages/frontend/src/graphql/mutations/createWord.ts
"use client";

import { useMutation } from "@apollo/client/react";
import { graphql } from "@/gql";
import type { CreateWordMutationVariables } from "@/gql/graphql";

/** 自作単語帳への単語の追加（本人のみ）。 */
export const CreateWordDocument = graphql(`
  mutation CreateWord($wordbookId: ID!, $question: String!, $answer: String!) {
    createWord(wordbookId: $wordbookId, question: $question, answer: $answer) {
      success
      errors {
        field
        message
      }
      word {
        id
        question
        answer
      }
    }
  }
`);

export type { CreateWordMutationVariables };

export function useCreateWordMutation() {
  return useMutation(CreateWordDocument);
}

File: packages/backend/app/graphql/mutations/create_word.rb
module Mutations
  class CreateWord < BaseMyWordMutation
    argument :wordbook_id, ID, required: true, description: "追加先の自作単語帳"
    argument :question, String, required: true
    argument :answer, String, required: true

    def resolve(wordbook_id:, question:, answer:)
      return failure(unauthorized_errors) unless current_user

      wordbook = current_user.wordbooks.personal.kept.find_by(id: wordbook_id)
      return failure(wordbook_not_found_errors) unless wordbook

      # user は Word の before_validation で単語帳の所有者から引き継がれる。
      word = wordbook.words.build(question: question, answer: answer)

      return failure(validation_errors(word)) unless word.valid?

      word.save!

      { success: true, errors: [], word: word }
    end

    private

    def wordbook_not_found_errors
      [ { field: "wordbookId", message: "追加先の単語帳が見つかりません" } ]
    end
  end
end
